File: task 1/Task1_Ganesh.java
import java.io.*;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;

/*
 * Should not use 
 * javax.xml.* or 
 * org.w3c.* or 
 * org.jsoup.* or 
 * java.util.stream.* or 
 * java.util.regex.* or
 * html dom element libraries or
 * Pattern, matcher or any
 * other such libraries 
 * You should find the logic from scratch
 */

public class Task1_Ganesh {

	public static ArrayList<String> arraylist(String s) {

		StringBuffer sb = new StringBuffer(s);
		ArrayList<String> s1 = new ArrayList<String>();

		int i = 0;
		while (true) {
			String d = sb.substring(sb.indexOf("<"), sb.indexOf(">") + 1);
			s1.add(d);
			// s.replace(d," ");
			sb.delete(sb.indexOf("<"), sb.indexOf(">") + 1);
			i++;
			if (sb.length() == 0) {
				break;
			}
		}

		return s1;
	}

	public static int[] oc(int id, String s) {
		int open = 1, close = 0, si = 0, ei = 0;

		ArrayList<String> s1 = arraylist(s);

		for (String z : s1) {
			if (z.contains(String.format("\"%d", id))) {
				if (!z.contains("/>")) {
					si = s1.indexOf(z);
					break;
				}
			}
		}
		// System.out.println(si);

		for (int j = si + 1; true; j++) {
			if (s1.get(j).contains("id") && !s1.get(j).contains("/>")) {
				open += 1;
			}
			if (s1.get(j).contains("</")) {
				close += 1;
			}
			if (open == close) {
				ei += j;
				break;
			}

		}
		// System.out.println(ei);
		int[] arr = { si, ei };
		return arr;

	}

	public static int oc(int id, String s, boolean b) {
		int si = 0;

		ArrayList<String> s1 = arraylist(s);
		if (b) {
			for (String z : s1) {
				if (z.contains(String.format("\"%d", id))) {
					if (z.contains("/>")) {
						si = s1.indexOf(z);
						break;
					}
				}
			}
		}
		return si;
	}

	public static void main(String[] args) throws IOException {

		String givenString = new String(Files.readAllBytes(Paths.get("htmlFile.txt")));
		givenString = givenString.replaceAll("\\t", "");
		givenString = givenString.replaceAll("\\r", "");
		givenString = givenString.replaceAll("\\n", "");

		System.out.println("Given String:");
		System.out.println(givenString);

		System.out.println("\nOutput for append:");
		System.out.println(append(givenString, "6", "<img id=\"appendSample2\" />"));

		System.out.println("\nOutput for prepend:");
		System.out.println(prepend(givenString, "10", "<br id=\"prependSample2\" />"));

		System.out.println("\nOutput for after:");
		System.out.println(after(givenString, "7", "<div id=\"afterSample2\"><div>"));

		System.out.println("\nOutput for before:");
		System.out.println(before(givenString, "13", "<span id=\"beforeSample2\"><span>"));

		// System.out.println(before(givenString,"9","*****"));
	}

	public static String append(String givenString, String id, String stringToAdd) {

		String outputString = "";
		int Id = Integer.valueOf(id);

		ArrayList<String> arrlist = arraylist(givenString);
		int[] arr = oc(Id, givenString);

		int start = arr[0];
		int end = arr[1];

		boolean single = false;

		for (String s : arrlist) {
			if (s.contains(String.format("\"%d", Id)) && s.contains("/>")) {
				single = true;
				outputString += "No Output";
				return outputString;
			}
		}
		if (!single) {
			for (int i = 0; i < arrlist.size(); i++) {
				if (i != end) {
					outputString += arrlist.get(i);
				}
				if (i == end) {
					outputString += stringToAdd;
					outputString += arrlist.get(i);
				}
			}
		}
		return outputString;
	}

	public static String prepend(String givenString, String id, String stringToAdd) {

		String outputString = "";
		int Id = Integer.valueOf(id);

		ArrayList<String> arrlist = arraylist(givenString);
		int[] arr = oc(Id, givenString);

		int start = arr[0];
		int end = arr[1];
		boolean single = false;

		for (String s : arrlist) {
			if (s.contains(String.format("\"%d", Id)) && s.contains("/>")) {
				single = true;
				outputString += "No Output";
				return outputString;

			}
		}

		if (!single) {
			for (int i = 0; i < arrlist.size(); i++) {
				if (i != start) {
					outputString += arrlist.get(i);
				}
				if (i == start) {
					outputString += arrlist.get(i);
					outputString += stringToAdd;
				}
			}
		}
		return outputString;
	}

	public static String after(String givenString, String id, String stringToAdd) {

		String outputString = "";
		int Id = Integer.valueOf(id);

		ArrayList<String> arrlist = arraylist(givenString);

		boolean single = false;

		for (String s : arrlist) {
			if (s.contains(String.format("\"%d", Id)) && s.contains("/>")) {
				single = true;
				break;
			}
		}

		if (single) {
			int index = oc(Id, givenString, single);
			for (int i = 0; i < arrlist.size(); i++) {
				if (i != index) {
					outputString += arrlist.get(i);
				}
				if (i == index) {
					outputString += arrlist.get(i);
					outputString += stringToAdd;
				}
			}
			return outputString;
		} else {
			int[] arr = oc(Id, givenString);

			int end = arr[1];

			for (int i = 0; i < arrlist.size(); i++) {
				if (i != end) {
					outputString += arrlist.get(i);
				}
				if (i == end) {
					outputString += arrlist.get(i);
					outputString += stringToAdd;
				}
			}

			return outputString;
		}
	}

	public static String before(String givenString, String id, String stringToAdd) {

		String outputString = "";
		int Id = Integer.valueOf(id);

		ArrayList<String> arrlist = arraylist(givenString);

		boolean single = false;

		for (String s : arrlist) {
			if (s.contains(String.format("\"%d", Id)) && s.contains("/>")) {
				single = true;
				break;
			}
		}

		if (single) {
			int index = oc(Id, givenString, single);
			for (int i = 0; i < arrlist.size(); i++) {
				if (i != index) {
					outputString += arrlist.get(i);
				}
				if (i == index) {
					outputString += stringToAdd;
					outputString += arrlist.get(i);
				}
			}
			return outputString;
		} else {
			int[] arr = oc(Id, givenString);

			int start = arr[0];

			for (int i = 0; i < arrlist.size(); i++) {
				if (i != start) {
					outputString += arrlist.get(i);
				}
				if (i == start) {
					outputString += stringToAdd;
					outputString += arrlist.get(i);
				}
			}

			return outputString;
		}
	}
}

File: task 1/task1_Ganesh.py
s = open("htmlFile.txt")
s = s.readlines()
l = list()
givenString = ""
for i in s:
    l.append(i.strip("\n").strip("\t").strip("\r"))
for i in l:
    givenString += i


def oc(id):
    ope, clos, si, ei = 1, 0, 0, 0

    for i in l:
        if("\""+str(id)+"\"" in i):
            if (not "/>" in i):
                si = l.index(i)
                break

    j = si+1
    while True:
        if("id" in l[j] and not "/>" in l[j]):
            ope += 1
        if("</" in l[j]):
            clos += 1
        if(ope == clos):
            ei += j
            break
        j += 1
    return [si, ei]


def oc1(id, single):
    si = 1

    if (single):
        for i in l:
            if ("\""+str(id)+"\"" in i):
                if ("/>" in i):
                    si = l.index(i)
                    break
    return si


def append(gs, id, re):
    outputString = ""
    id = int(id)
    arr = oc(id)

    start = arr[0]
    end = arr[1]
    single=False
    
    for i in l:
        if("\""+str(id)+"\"" in i and "/>" in i):
            single = True
            outputString += "No Output"
            return outputString

  
    if(not single):
        for i in range(len(l)):
            if(i != end):
                outputString += l[i]
            if (i == end):
                outputString += re
                outputString += l[i]
    return outputString


def prepend(gs, id, re):
    outputString = ""
    id = int(id)
    arr = oc(id)
    single=False

    start = arr[0]
    end = arr[1]

    for i in l:
        if("\""+str(id)+"\"" in i and "/>" in i):
            single = True
            outputString += "No Output"
            return outputString
    if(not single):
        for i in range(len(l)):
            if(i != start):
                outputString += l[i]
            if (i == start):
                outputString += l[i]
                outputString += re

    return outputString


def after(gs, id, re):
    outputString = ""
    id = int(id)

    single = False

    for i in l:
        if("\""+str(id)+"\"" in i and "/>" in i):
            single = True
            break

    if(single):
        index = oc1(id, single)
        for j in range(len(l)):
            if(j != index):
                outputString += l[j]
            if(j == index):
                outputString += l[j]
                outputString += re
        return outputString

    else:
        arr = oc(id)
        start = arr[0]
        end = arr[1]

        for j in range(len(l)):
            if(j != end):
                outputString += l[j]
            if(j == end):
                outputString += l[j]
                outputString += re
        return outputString


def before(gs, id, re):
    outputString = ""
    id = int(id)

    single = False

    for i in l:
        if("\""+str(id)+"\"" in i and "/>" in i):
            single = True
            break

    if(single):
        index = oc1(id, single)
        for j in range(len(l)):
            if(j != index):
                outputString += l[j]
            if(j == index):
                outputString += re
                outputString += l[j]

        return outputString

    else:
        arr = oc(id)
        start = arr[0]
        end = arr[1]

        for j in range(len(l)):
            if(j != start):
                outputString += l[j]
            if(j == start):
                outputString += re
                outputString += l[j]

        return outputString


if __name__ == "__main__":
    print("Given String:")
    print(givenString)

    print("\nOutput for append:")
    print(append(givenString, "6", "<img id=\"appendSample2\" />"))
    print("\nOutput for prepand:")
    print(prepend(givenString, "10", "<br id=\"prependSample2\" />"))
    print("\nOutput for after:")
    print(after(givenString, "7", "<div id=\"afterSample2\"><div>"))
    print("\nOutput for before:")
    print(before(givenString, "13", "<span id=\"beforeSample2\"><span>"))
  
  
  